### PR TITLE
#580: Handle devtools-background connection errors

### PR DESCRIPTION
**Pull request by fregante**
While working on #621 I noticed that these errors are _probably_ not caught.

Next:

- [x] Test this code

### CODE REVIEW

File: src/chrome.ts [hunk 1]
```diff
@@ -98,6 +98,25 @@ export function getChromeExtensionId(): string {
...
+    port.onDisconnect.addListener(() => {
+      // TODO: Also handle port.error in Firefox https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/Port#type
+      const message =
+        chrome.runtime.lastError?.message ??
+        "There was an error while connecting to the runtime";
+      reject(new Error(message));
+    });
+
+    // If onDisconnect hasn't been called by now, we assume the connection succeeded. There's no port.onConnect
+    setTimeout(resolve, 0, port);
```
Comment (fregante) [inline src/chrome.ts:116]: The code hasn't been tested directly, it's based on this:

Screen Shot 10

- `sync lastError undefined` shows that the lastError that the previous code was looking for was not set in time
- ignore _16_
- `onDisconnect lastError {}` shows that it's available in there (the object actually expands into `{message: "Could not establish connection. Receiving end does not exist."}`)
- `Connected now…` shows that `onDisconnect` is called within the first possible timeout, i.e. almost immediately

File: src/background/devtools/external.ts [hunk 2]
```diff
@@ -94,22 +94,18 @@ export async function callBackground(
...
-    if (browser.runtime.lastError) {
-      throw new Error(
-        `Error sending devtools notification: ${browser.runtime.lastError.message}`
-      );
-    } else {
-      return;
+    try {
+      port.postMessage(message);
+    } catch (error) {
+      throw new Error(`Error sending devtools notification: ${error.message}`);
```
Comment (fregante) [inline src/background/devtools/external.ts:100]: `postMessage` does not seem to set `lastError` at all, but it just throws with a proper error, as seen in:

Screen Shot 11

File: src/devTools/protocol.ts [hunk 3]
```diff
@@ -33,22 +34,16 @@ export async function connectDevtools(): Promise<Runtime.Port> {
...
     console.error("Devtools cannot connect to the background page", {
-      error: browser.runtime.lastError,
+      error,
     });
-    throw new Error(browser.runtime.lastError.message);
-  }
-
-  if (!port) {
-    throw new Error("Could not get connection port to background page");
+    throw error;
```
Comment (fregante) [inline src/devTools/protocol.ts:46]: It seems that `port` is always returned in order to attach the `onDisconnect` listener. Let me know if you found a way to not return it at all.

File: src/background/devtools/external.ts (hunk 2 above)
Comment (fregante) [reply src/background/devtools/external.ts:100]: I'm unable to test this part in PB because as soon as I refresh the extension or manually `disconnect()` the port in the background, a lot of error pop up in various parts of the extension, from `hasListeners` to `webext-detect-page`, never reading `postMessage` at all.

That actually brought me to reduce the chance of `webext-detect-page` throwing errors by adding caching in the next version: https://github.com/fregante/webext-detect-page/commit/1cc2e123c32b90fb849b2f922a3de5b456757d83

File: src/background/devtools/external.ts [hunk 4]
```diff
@@ -94,22 +94,18 @@ export async function callBackground(
...
     }
   } else {
     return new Promise((resolve, reject) => {
       devtoolsHandlers.set(nonce, [resolve, reject]);
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/background/devtools/external.ts:104]: Thant's correct

For the Firefox external messaging, I do the same thing (for both sides): http://github.com/pixiebrix/pixiebrix-extension/blob/463004f75562d12e826348f3a816b594b129ff36/src/contentScript/externalProtocol.ts#L40-L40. 

Not sure why, but for the external page I made separate maps for the fulfilled and rejected callbacks

File: src/chrome.ts (hunk 1 above)
Comment (twschiller) [reply src/chrome.ts:116]: This method doesn't work? https://developer.chrome.com/docs/extensions/reference/runtime#event-onConnect

If it only gets triggered from the other side of the port, you could immediately send back a greeting message, so the initiator knows the port is now working

File: src/devTools/protocol.ts (hunk 3 above)
Comment (twschiller) [reply src/devTools/protocol.ts:46]: Not sure I follow 

If you look at usages of `connectDevtools` it's used in the React hook `useDevConnection`, which is currently used by the devtools React front-end. The port is passed to child components via React's Context: https://reactjs.org/docs/context.html feature

**Review by twschiller — CHANGES_REQUESTED** [review]
Use onConnect to make connection detection more reliable: https://github.com/pixiebrix/pixiebrix-extension/pull/623#discussion_r659416220

File: src/devTools/protocol.ts (hunk 3 above)
Comment (fregante) [reply src/devTools/protocol.ts:46]: Sorry this comment was probably on the wrong line. I was referring to this piece of code that this PR is deleting:

```js
if (!port) {
	throw new Error("Could not get connection port to background page");
}
```

From what I can tell `chrome.runtime.connect()` _always_ returns `port`. I was asking you if you found a way that causes `connect()` to return undefined instead.

File: src/background/devtools/internal.ts [hunk 5]
```diff
@@ -237,6 +237,9 @@ function connectDevtools(port: Runtime.Port): void {
       }`
     );
 
+    // ping back
+    port.postMessage("Port open");
```
Comment (twschiller) [inline src/background/devtools/internal.ts:241]: I think this should be following our message structure of `{type: string, payload: string}` to simplify the kinds of messages our code expects to handle

Also, an additional comment on why the "ping back" is necessary would be helpful for future readers

File: src/chrome.ts [hunk 6]
```diff
@@ -96,6 +98,36 @@ export function getChromeExtensionId(): string {
   return isEmpty(manualKey ?? "") ? CHROME_EXTENSION_ID : manualKey;
 }
 
+/** Connect to the extension runtime and throw proper errors if it fails. NOTE: It expects a message from the other side */
+export async function runtimeConnect(name: string): Promise<Runtime.Port> {
+  if (isBackgroundPage()) {
+    throw new Error("runtimeConnect cannot be called from the background page");
+  }
+
+  const deferred = pDefer();
```
Comment (twschiller) [inline src/chrome.ts:107]: Chose a better name for this variable, e.g., `deferredConnectionPromise` so it's more clear what the use of the variable is. 

It's hard to follow that this promise is what actually gets waited on before the returned value

File: src/chrome.ts [hunk 7]
```diff
@@ -96,6 +98,36 @@ export function getChromeExtensionId(): string {
   return isEmpty(manualKey ?? "") ? CHROME_EXTENSION_ID : manualKey;
 }
 
+/** Connect to the extension runtime and throw proper errors if it fails. NOTE: It expects a message from the other side */
```
Comment (twschiller) [inline src/chrome.ts:101]: Explain why it "expects a message from the other side". That is, it expects a message from the other side in order to determine that the port connection was successfully opened

File: src/chrome.ts [hunk 8]
```diff
@@ -96,6 +98,36 @@ export function getChromeExtensionId(): string {
...
+      "There was an error while connecting to the runtime";
+    deferred.reject(new Error(message));
+  };
+
+  const port = browser.runtime.connect(null, { name });
+  port.onMessage.addListener(deferred.resolve); // Any message is fine
+  port.onDisconnect.addListener(onDisconnect);
+
+  try {
+    await pTimeout(deferred.promise, 1000);
```
Comment (twschiller) [inline src/chrome.ts:123]: Add a comment explaining why a timeout is necessary here? Is it because onDisconnect doesn't get called if the other end doesn't respond?

File: src/devTools/protocol.ts (hunk 3 above)
Comment (twschiller) [reply src/devTools/protocol.ts:46]: Not that I recall, I think I was just being safe/overcautious while trying to debug a port issue at some point. Feel free to remove

**Review by twschiller — APPROVED** [review]

**Comment by twschiller** [comment]
@fregante This looks good. I added some suggestions on how to improve the readability of the code for people who come across it in the future. This stuff is hairy, so I'd like to err on the side of over explaining why certain things are there

Go ahead and merge when ready

File: src/chrome.ts (hunk 8 above)
Comment (fregante) [reply src/chrome.ts]: Correct. It's only here to catch developer mistakes of not answering the message during `onConnect`. Assuming that the background page is always available, this should only happen within a milliseconds. Sometimes 1, sometimes 90.

File: src/chrome.ts (hunk 6 above)
Comment (fregante) [reply src/chrome.ts]: Dropped the variable, I left the normal resolve/reject. 

Deferreds are to be read as "open promises" where the promise and resolve/reject callbacks exist in the same scope. I used this to allow removing the listeners in `finally`
